scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/pskstore/InMemoryPskStore.java

```diff
@@ -21,56 +22,46 @@
 import java.util.Map;
 import java.util.concurrent.ConcurrentHashMap;
 
-import org.eclipse.californium.scandium.dtls.ServerNameResolver;
 import org.eclipse.californium.scandium.util.ServerName;
 import org.eclipse.californium.scandium.util.ServerNames;
+import org.eclipse.californium.scandium.util.ServerName.NameType;
 
 /**
- * An in-memory pre-shared-key storage. 
+ * An in-memory pre-shared key storage.
+ * <p>
  * If you don't need to initiate connection,
  * you could just add identity/key with {@link #setKey(String, byte[])}.
  * If you need to initiate connection, 
  * you should add known peers with {@link #addKnownPeer(InetSocketAddress, String, byte[])}.
- * 
+ * <p>
  * To be used only for testing and evaluation. 
- * You are supposed to store your key in a secure way: 
+ * You are supposed to store your key in a secure way:
  * keeping them in-memory is not a good idea.
  */
-public class InMemoryPskStore implements PskStore, ServerNameResolver {
+public class InMemoryPskStore implements PskStore {
+
+	private static final ServerName GLOBAL_SCOPE = ServerName.from(NameType.UNDEFINED, new byte[0]);
 
 	private final Map<ServerName, Map<String, byte[]>> scopedKeys = new ConcurrentHashMap<>();
-	private final Map<String, byte[]> keys = new ConcurrentHashMap<>();
-	private final Map<InetSocketAddress, String> identitiesByAddress = new ConcurrentHashMap<>();
-	private final Map<InetSocketAddress, ServerNames> serverNamesByAddress = new ConcurrentHashMap<>();
+	private final Map<InetSocketAddress, Map<ServerName, String>> scopedIdentities = new ConcurrentHashMap<>();
 
 	@Override
 	public byte[] getKey(final String identity) {
 
 		if (identity == null) {
 			throw new NullPointerException("identity must not be null");
 		} else {
-			return getKeyFromMap(identity, keys);
-		}
-	}
-
-	private static byte[] getKeyFromMap(final String identity, final Map<String, byte[]> keyMap) {
-
-		byte[] result = null;
-		if (keyMap != null) {
-			byte[] key = keyMap.get(identity);
-			if (key != null) {
-				// defensive copy
-				result = Arrays.copyOf(key, key.length);
+			synchronized (scopedKeys) {
```

Comment:
I'm not sure, if `addKnownPeer` should be called dynamically during usage/handshakes. I would guess, that, if this `PskStore` is intended to be used dynamically during execution time, a `removeKnownPeer` is also required. If a dynamic usage should be supported, I would prefer to synchronize on `this`.

Reply:
by *dynamic*, I suppose, you mean adding and removing peers during runtime, i.e. after the connector has been started, right? I didn't really think about that, since the *addKnownPeer* method had been there before already. However, adding a *removeKnownPeer* method seem reasonable to me. But why do you think it would be better to synchronize on the class instance itself?

Reply:
Because the synchronization belongs to both collections. Otherwise we may add a comment, that accessing the other collection should also be protected by synchronization the other. Therefore it seems for me to be more straight to synchronize on `this`.

Reply:
I would rather add the comment and synchronize on an internal object instead of the class instance itself. Otherwise other code can interfere with the synchronization simply by also synchronizing on the class instance as well .